openscm/adapters/__init__.py

```diff
@@ -29,72 +31,119 @@ class Adapter(metaclass=ABCMeta):
     _current_time: np.datetime64
     """Current time when using :func:`step`"""
 
-    _initialized: bool
-    """``True`` if model has been initialized via :func:`_initialize_model`"""
+    _parameters: ParameterSet
+    """Input parameter set"""
 
     _output: ParameterSet
     """Output parameter set"""
 
-    _parameters: ParameterSet
-    """Input parameter set"""
+    _parameter_views: Dict[HierarchicalName, ParameterInfo]
+    """Parameter views available with this model"""
+
+    _parameter_versions: Dict[HierarchicalName, int]
+    """Parameter versions (last time parameter was passed to the model)"""
+
+    _openscm_standard_parameter_mappings: Dict[HierarchicalName, str]
+    """
+    Mapping from OpenSCM parameters to model parameters.
+
+    If required, use property setters to add extra behaviour (like calculating a model
+    parameter based on the value of the OpenSCM parameter) when setting a model
+    parameter from an OpenSCM parameter.
+    """
+
+    _direct_set: bool
+    """Be careful about overwriting other parameters when setting a model parameter?"""
 
     def __init__(self, input_parameters: ParameterSet, output_parameters: ParameterSet):
         """
-        Initialize.
+        Initialize the adapter as well as the model sitting underneath it.
+
+        *Note:* as part of this process, all available model parameters are added to
+        ``input_parameters`` (if they're not already there).
 
         Parameters
         ----------
         input_parameters
             Input parameter set to use
+
         output_parameters
             Output parameter set to use
         """
         self._parameters = input_parameters
         self._output = output_parameters
-        self._initialized = False
-        self._initialized_inputs = False
-        self._current_time = 0
+        self._parameter_views = {}
+        self._parameter_versions = {}
+        self._direct_set = False
+        self._initialize_model()
```

Comment:
Please revert back to the former model with the separate initialize methods. That is to

- separate initialization for the model itself as well as for a single run
- and for those ****** models that use file I/O so the `__init__` function does not get hold up by those file initializations

Reply:
To try and explain why I made the change more clearly: if you keep track of the views' versions then you can do everything in `reset`. The reason is that you only update things when you have to which means you can delay all operations as late as possible. If we wanted, we could shift initialise model into `reset` with some flag to make sure it's only done once. 

> * for those ****** models that use file I/O so the `__init__` function does not get hold up by those file initializations

Is this desirable? As the user I feel like I would want my model to be spun up (whatever that means) so that when I interact with it thereafter it's a) fast (cause it's not doing as many io steps) and b) definitely up to date with my adapter. I'm not super fussed either way but this was my thinking when I put `_initialise_model` in `__init__`.